test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs

```diff
@@ -6613,5 +6613,231 @@ public void SummarizeEntryPointTest()
 
             Done();
         }
+
+        [Fact]
+        void RankingWithColumnIdEntryPoint()
+        {
+            Env.ComponentCatalog.RegisterAssembly(typeof(LightGbmBinaryModelParameters).Assembly);
```

Comment:
I added this test with pretty much the same graph produced by the NimbusML test that created this issue.

For some reason it seems that to test LightGBM trainers using entrypoints we actually need to explicitly register the LightGbm assembly... otherwise the LightGBM's entrypoints aren't found when running the test and an exception is thrown. Does anybody know why we need to have this special treatment for LightGBM on entrypoint tests? @yaeldekel @ganik ? I was confused by this, and it might become relevant to know why this is necessary in the future. Thanks !

Reply:
Hi @antoniovs1029 , The reason you need to register the LightGbm assembly when you use entrypoints, is because the graph runner gets the graph as a json object, and from the entrypoint name it has to figure out which method to call. This is done using the appropriate LoadableClassAttribute (or EntrypointModuleAttribute, I forget exactly which one is used), which is defined in the LightGbm assembly.
This is similar to the reason we have to register the required assemblies when running maml, since it has to figure out the correct trainer/estimator to instantiate from the string specified in the commandline, and also why we have to register assemblies when loading a model from disk.
Specifically in the case of your test, I think that in the constructor of one of the base classes of the test project some assemblies are registered, if LightGbm is not one of them you can add it there.

Reply:
Thanks for your reply. I understand that if I don't register the assembly, then when calling [ComponentCatalog.TryFindEntryPoint()](https://github.com/dotnet/machinelearning/blob/3d11ce42ab108d467ea6076dc14eaded57f46172/src/Microsoft.ML.Data/EntryPoints/EntryPointNode.cs#L483-L484) it won't be able to find `Trainers.LightGbmRanker` because it hasn't be registered. What I still don't understand is why we need to explicitly register LightGBM assembly, but not others... for instance, `FastTree` trainers are registered even if I don't register them explicitly in the test. So I guess my question could also be phrased as: Which trainers get always registered and where does this happen?

---
In reply to: [441779796](https://github.com/dotnet/machinelearning/pull/5227#discussion_r441779796) [](ancestors = 441779796)

Reply:
Ok, so simply whenever we run tests, most of the assemblies are registered here:
[https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/test/Microsoft.ML.TestFramework/EnvironmentExtensions.cs#L20-L27](https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/test/Microsoft.ML.TestFramework/EnvironmentExtensions.cs#L20-L27)

Which, for whatever reason, doesn't include LightGBM or these other assemblies, which need to be registered manually when working with Entrypoints:
[https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs#L328-L335](https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs#L328-L335)

---
In reply to: [441993836](https://github.com/dotnet/machinelearning/pull/5227#discussion_r441993836) [](ancestors = 441993836,441779796)

Reply:
I was confused because I didn't understand/remember why typically users don't need to manually register assemblies, when loading models from disk, but I think it's simply because whenever a model is loaded the `EnsuredLoaderAssemblyIsRegistered` method is called, which actually tries to register the assembly that is needed

[https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/src/Microsoft.ML.Core/Data/ModelLoading.cs#L284-L291](https://github.com/dotnet/machinelearning/blob/bb13d629000c218136e741b643767cf45ae12fc4/src/Microsoft.ML.Core/Data/ModelLoading.cs#L284-L291)

Anyway. Everything clear now.

---
In reply to: [444673165](https://github.com/dotnet/machinelearning/pull/5227#discussion_r444673165) [](ancestors = 444673165,441993836,441779796)